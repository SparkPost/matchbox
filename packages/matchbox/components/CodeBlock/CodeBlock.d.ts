export default CodeBlock;
declare function CodeBlock(props: any): any;
declare namespace CodeBlock {
    namespace propTypes {
        const code: any;
        const className: any;
        const height: any;
        const numbered: any;
        const dark: any;
    }
    namespace defaultProps {
        const numbered_1: boolean;
        export { numbered_1 as numbered };
        const dark_1: boolean;
        export { dark_1 as dark };
    }
}
