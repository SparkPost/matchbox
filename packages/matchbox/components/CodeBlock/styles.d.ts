export function pre(props: any): string;
export function code(props: any): string;
export function line(props: any): string;
export function prefix(): string;
export function chevron(): string;
