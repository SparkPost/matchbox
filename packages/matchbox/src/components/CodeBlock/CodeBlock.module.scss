@import '../../styles/config';

.CodeBlock {
  position: relative;
  font-family: font-family(monospace);
  color: color(gray, 2);
  background: color(gray, 9);
  font-size: rem(13);
  line-height: 21px;
}

.Code {
  padding: spacing() spacing() spacing() rem(50);
  position: absolute;
}

.PrefixWrapper {
  position: absolute;
  top: spacing();
  left: rem(15);
}

.LineChevron {
  display: block;
  margin: 0;
  height: 21px;
  fill: color(gray, 6);
}

.LineNumber {
  display: block;
  text-align: right;
  margin: 0;
  height: 21px;
  font-size: rem(12);
  color: color(gray, 6);
}
