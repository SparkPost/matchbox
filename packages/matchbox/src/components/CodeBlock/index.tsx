export { default as CodeBlock } from './CodeBlock';
export * from './CodeBlock';
