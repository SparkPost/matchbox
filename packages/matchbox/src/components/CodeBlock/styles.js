import { tokens } from '@sparkpost/design-tokens';

const lineHeight = '25px';
const fontSize = tokens.fontSize_200;

export const pre = props => `
  position: relative;
  display: grid;
  grid-template-columns: ${tokens.spacing_700} auto;
  font-family: ${tokens.fontFamily_monospace};
  border-radius: ${tokens.borderRadius_100}
  background-color: ${props.dark ? tokens.color_gray_900 : tokens.color_gray_100};
  border: 1px solid ${tokens.color_gray_400};
  overflow: scroll;
  padding: ${tokens.spacing_600} ${tokens.spacing_400} ${tokens.spacing_600} 0;
`;

export const code = props => `
  position: absolute;
  padding: ${tokens.spacing_600} ${tokens.spacing_600} ${tokens.spacing_600} ${tokens.spacing_800};
  color: ${props.dark ? tokens.color_white : tokens.color_gray_800};
  font-size: ${fontSize};
  line-height: ${lineHeight};
`;

export const line = props => `
  display: block;
  text-align: right;
  line-height: ${lineHeight};
  font-size: ${fontSize};
  color: ${props.dark ? tokens.color_white : tokens.color_gray_700};
  user-select: none;
`;

export const prefix = () => `
  padding-right: ${tokens.spacing_200};
`;

export const chevron = () => `
  width: ${tokens.fontSize_400};
  height: ${tokens.fontSize_400};
  fill: ${tokens.color_gray_500};
`;
