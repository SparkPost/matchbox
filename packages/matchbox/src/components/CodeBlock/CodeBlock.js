import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { ChevronRight } from '@sparkpost/matchbox-icons';
import { pre, code, prefix, line, chevron } from './styles';

const StyledPre = styled('pre')`
  ${pre}
`;

const StyledCode = styled('code')`
  ${code}
`;

const StyledCodePrefix = styled('div')`
  ${prefix}
`;

const StyledLineNumber = styled('span')`
  ${line}
`;

const StyledChevron = styled(ChevronRight)`
  ${chevron}
`;

function CodeBlock(props) {
  const { children, code, height, className, numbered, dark } = props;

  return (
    <StyledPre
      className={className}
      style={{ height: height ? `${height}px` : 'auto' }}
      dark={dark}
    >
      <CodePrefix dark={dark} code={code} numbered={numbered} />

      <StyledCode dark={dark}>
        {/* If children are passed in, render those, and pass in the code as a child of that element */}
        {children ? (
          React.Children.map(children, child => {
            return React.cloneElement(child, { children: code });
          })
        ) : (
          <>{code}</>
        )}
      </StyledCode>
    </StyledPre>
  );
}

function CodePrefix({ code, numbered, dark }) {
  const rows = code.split(/\r\n|\r|\n/);

  return (
    <StyledCodePrefix aria-hidden="true">
      {rows.map((_row, rowIndex) => {
        const key = `code-block-row-${rowIndex}`;

        if (numbered) {
          return (
            <StyledLineNumber dark={dark} key={key}>
              {rowIndex + 1} {/* Line numbers in code editors do not start with "0"*/}
            </StyledLineNumber>
          );
        }

        return (
          <StyledLineNumber dark={dark} key={key}>
            <StyledChevron />
          </StyledLineNumber>
        );
      })}
    </StyledCodePrefix>
  );
}

CodeBlock.propTypes = {
  /**
   * The string of code to render
   */
  code: PropTypes.string,
  className: PropTypes.string,
  /**
   * Height in pixels of the <pre> block
   */
  height: PropTypes.number,
  /**
   * Render line numbers
   */
  numbered: PropTypes.bool,
  /**
   * Whether the code block is styled dark or light
   */
  dark: PropTypes.bool,
};

CodeBlock.defaultProps = {
  numbered: false,
  dark: false,
};

export default CodeBlock;
