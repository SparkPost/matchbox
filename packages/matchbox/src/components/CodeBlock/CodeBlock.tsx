import React, { ReactNode } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { ChevronRight } from '@sparkpost/matchbox-icons';
import { pre, code, prefix, line, chevron } from './styles';

const StyledPre = styled('pre')`
  ${pre}
`;

const StyledCode = styled('code')`
  ${code}
`;

const StyledCodePrefix = styled('div')`
  ${prefix}
`;

const StyledLineNumber = styled('span')`
  ${line}
`;

const StyledChevron = styled(ChevronRight)`
  ${chevron}
`;

type CodeBlockProps = {
  /**
   * The string of code to render
   */
  code: string;
  className?: string;
  /**
   * Height in pixels of the <pre> block
   */
  height?: number;
  /**
   * Render line numbers
   */
  numbered?: boolean;
  /**
   * Whether the code block is styled dark or light
   */
  dark?: boolean;
  children?: React.ReactNode;
};

function CodeBlock(props: CodeBlockProps): JSX.Element {
  const { children, code, height, className, numbered = false, dark = false } = props;

  return (
    <StyledPre
      className={className}
      style={{ height: height ? `${height}px` : 'auto' }}
      dark={dark}
    >
      <CodePrefix dark={dark} code={code} numbered={numbered} />

      <StyledCode dark={dark}>
        {/* If children are passed in, render those, and pass in the code as a child of that element */}
        {children ? (
          React.Children.map<ReactNode, ReactNode>(children, (child) => {
            if (React.isValidElement(child)) {
              return React.cloneElement(child, { children: code });
            }
          })
        ) : (
          <>{code}</>
        )}
      </StyledCode>
    </StyledPre>
  );
}

function CodePrefix({ code, numbered, dark }) {
  const rows = code.split(/\r\n|\r|\n/);

  return (
    <StyledCodePrefix aria-hidden="true">
      {rows.map((_row, rowIndex) => {
        const key = `code-block-row-${rowIndex}`;

        if (numbered) {
          return (
            <StyledLineNumber dark={dark} key={key}>
              {rowIndex + 1} {/* Line numbers in code editors do not start with "0"*/}
            </StyledLineNumber>
          );
        }

        return (
          <StyledLineNumber dark={dark} key={key}>
            <StyledChevron />
          </StyledLineNumber>
        );
      })}
    </StyledCodePrefix>
  );
}

export default CodeBlock;
