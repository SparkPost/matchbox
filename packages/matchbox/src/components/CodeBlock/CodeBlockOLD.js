import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { ChevronRight } from '@sparkpost/matchbox-icons';

import styles from './CodeBlock.module.scss';

class CodeBlock extends Component {
  static displayName = 'CodeBlock';

  static defaultProps = {
    height: 355,
    numbered: false,
  };

  static propTypes = {
    /**
     * The string of code to render
     */
    code: PropTypes.string,

    /**
     * Height in pixels of the <pre> block
     */
    height: PropTypes.number,

    /**
     * Use line numbers instead of chevrons
     */
    numbered: PropTypes.bool,
  };

  renderPrefix = (row, i) => {
    if (this.props.numbered) {
      return (
        <div key={i} className={styles.LineNumber}>
          {i}
        </div>
      );
    } else {
      return <ChevronRight key={i} className={styles.LineChevron} />;
    }
  };

  render() {
    const { code, height, className } = this.props;

    let prefixMarkup = null;
    const rows = code.split(/\r\n|\r|\n/);

    if (rows.length) {
      prefixMarkup = <div className={styles.PrefixWrapper}>{rows.map(this.renderPrefix)}</div>;
    }

    return (
      <pre className={classnames(styles.CodeBlock, className)} style={{ height: `${height}px` }}>
        <code className={styles.Code}>{code}</code>
        {prefixMarkup}
      </pre>
    );
  }
}

export default CodeBlock;
