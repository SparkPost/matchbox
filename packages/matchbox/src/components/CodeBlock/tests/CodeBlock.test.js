import React from 'react';
import CodeBlock from '../CodeBlock';
import { shallow } from 'enzyme';

describe('CodeBlock', () => {
  let wrapper;
  let props;
  beforeEach(() => {
    props = {
      code: 'curl -X POST\nhttps://api.sparkpost.com/api/v1/tranâ€¦\n'
    };

    wrapper = shallow(<CodeBlock {...props}/>);
  });

  it('renders with default props', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders line numbers', () => {
    wrapper.setProps({ numbered: true, height: 200 });
    expect(wrapper.find('.PrefixWrapper')).toMatchSnapshot();
  });

});
