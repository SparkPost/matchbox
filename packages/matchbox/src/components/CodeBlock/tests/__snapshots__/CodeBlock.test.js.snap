// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeBlock renders line numbers 1`] = `
<pre
  className="CodeBlock"
  style={
    Object {
      "height": "200px",
    }
  }
>
  <code
    className="Code"
  >
    curl -X POST
https://api.sparkpost.com/api/v1/tran…

  </code>
  <div
    className="PrefixWrapper"
  >
    <div
      className="LineNumber"
      key="0"
    >
      0
    </div>
    <div
      className="LineNumber"
      key="1"
    >
      1
    </div>
    <div
      className="LineNumber"
      key="2"
    >
      2
    </div>
  </div>
</pre>
`;

exports[`CodeBlock renders with default props 1`] = `
<pre
  className="CodeBlock"
  style={
    Object {
      "height": "355px",
    }
  }
>
  <code
    className="Code"
  >
    curl -X POST
https://api.sparkpost.com/api/v1/tran…

  </code>
  <div
    className="PrefixWrapper"
  >
    <ChevronRight
      className="LineChevron"
      key="0"
    />
    <ChevronRight
      className="LineChevron"
      key="1"
    />
    <ChevronRight
      className="LineChevron"
      key="2"
    />
  </div>
</pre>
`;
